### PR TITLE
feat: Add jax backend

### LINKED ISSUE #416: Investigate use of jax
# Description

As a compliment/replacement to Issue #79, consider using [JAX: Autograd and XLA](https://github.com/google/jax). This should be investigated and some examples made up.

### PR DESCRIPTION
**Pull request by lukasheinrich**
# Description

Resolves #416 

pdf evaluation works modulo this diff

```
diff --git a/pyhf/pdf.py b/pyhf/pdf.py
index 259bf44..4f299ca 100644
--- a/pyhf/pdf.py
+++ b/pyhf/pdf.py
@@ -423,7 +423,8 @@ class Model(object):
         tensorlib, _ = get_backend()
         lambdas_data = self.expected_actualdata(pars)
         summands = tensorlib.poisson_logpdf(maindata, lambdas_data)
-        tosum = tensorlib.boolean_mask(summands, tensorlib.isfinite(summands))
+        # tosum = tensorlib.boolean_mask(summands, tensorlib.isfinite(summands))
+        tosum = summands
         mainpdf = tensorlib.sum(tosum)
         return mainpdf
```

since bool-based indexing  seems to not work

```
import pyhf
import numpy as onp

pyhf.set_backend(pyhf.tensor.jax_backend())

m = pyhf.simplemodels.hepdata_like(signal_data = [5.], bkg_data = [17.], bkg_uncerts = [2.])
testdata = m.expected_data(m.config.suggested_init())
testpars = m.config.suggested_init()
value = m.logpdf(testpars,testdata)

print(type(value))
print(onp.asarray(value))
```

locally this gives

```
/Users/lukas/Code/pyhfdev/venv/lib/python3.6/site-packages/jax/lib/xla_bridge.py:143: UserWarning: No GPU found, falling back to CPU.
  warnings.warn('No GPU found, falling back to CPU.')
<class 'jax.interpreters.xla.DeviceArray'>
[-5.528393]
```

https://github.com/google/jax/issues/166

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR

```
* add JAX backend
```

### CODE REVIEW

**Comment by matthewfeickert** [comment]
I rebased this against `master` now that PR #379 and PR #380 are in.

**Comment by lukasheinrich** [comment]
@matthewfeickert any idea what's the issue with the CI here?

**Comment by kratsg** [comment]
> @matthewfeickert any idea what's the issue with the CI here?

This is the editable mode issue where we weren't following the pyproject.toml correctly. It looks like this PR needs to be rebased to include those fixes to the CI. I'll do that.

**Comment by kratsg** [comment]
See #455 .

**Comment by matthewfeickert** [comment]
@lukasheinrich What's the issue now? Or were you asking about what I [mentioned back in January](https://github.com/diana-hep/pyhf/pull/377#issuecomment-455044833)?

**Comment by matthewfeickert** [comment]
> This is the editable mode issue where we weren't following the pyproject.toml correctly.

Was it? I thought I responded to that `pip` issue as soon as it came up. My comment from January predates them.

File: pyhf/tensor/jax_backend.py [hunk 1]
```diff
@@ -13,7 +14,7 @@ class jax_backend(object):
     def __init__(self, **kwargs):
         self.name = 'jax'
 
-    def clip(self, tensor_in, min, max):
+    def clip(self, tensor_in, min_value, max_value):
```
Comment (matthewfeickert) [inline pyhf/tensor/jax_backend.py]: @kratsg This isn't strictly necessary, as I have control over the codefactor checker so I just turned them off for this PR. So if you don't want to have this backen's functions be slightly different then you can revert/drop this.

Comment (kratsg) [reply pyhf/tensor/jax_backend.py]: I changed all backends to match the signature. I think it's probably cleaner if we do it this way...

Comment (matthewfeickert) [reply pyhf/tensor/jax_backend.py]: Okay. I'm fine with it.

File: pyhf/tensor/jax_backend.py [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+
+    def gather(self, tensor, indices):
+        return tensor[indices]
+
+    def boolean_mask(self, tensor, mask):
+        return tensor[mask]
+
+    def isfinite(self, tensor):
+        return np.ones_like(tensor, dtype=onp.bool)
+        # return np.isfinite(tensor)
```
Comment (matthewfeickert) [inline pyhf/tensor/jax_backend.py]: Looks like there are some commented out bits that need to get cleaned up or given further information

Comment (kratsg) [reply pyhf/tensor/jax_backend.py]: these were copied over from numpy_backend.py

**Comment by kratsg** [comment]
`jax` isn't working atm:

```
(Pdb) test_statistic
array([[22.64193765],
       [22.64099121],
       [22.65319824],
       [ 0.        ]])
(Pdb) q_mu
DeviceArray([0.], dtype=float32)
```

`tests/test_backend_consistency.py` has `jax` returning `0` instead for q_mu here (for normal-50)

**Comment by kratsg** [comment]
Running the example in the PR description in this branch as-is (no changes to `pyhf/pdf.py` for the boolean indexing) works:

```
Python 3.7.3 (default, Mar 27 2019, 09:23:32) 
[Clang 9.0.0 (clang-900.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import pyhf
>>> import numpy as onp
>>> 
>>> pyhf.set_backend(pyhf.tensor.jax_backend())
>>> 
>>> m = pyhf.simplemodels.hepdata_like(signal_data = [5.], bkg_data = [17.], bkg_uncerts = [2.])
/Users/kratsg/.virtualenvs/pyhf/lib/python3.7/site-packages/jax/lib/xla_bridge.py:130: UserWarning: No GPU/TPU found, falling back to CPU.
  warnings.warn('No GPU/TPU found, falling back to CPU.')
>>> testdata = m.expected_data(m.config.suggested_init())
>>> testpars = m.config.suggested_init()
>>> value = m.logpdf(testpars,testdata)
>>> 
>>> print(type(value))
<class 'jax.interpreters.xla.DeviceArray'>
>>> print(onp.asarray(value))
[-5.528393]
>>> exit()
```

so now I need to understand how the tests are failing -- or perhaps they're not anymore after fixing in #491.

**Comment by kratsg** [comment]
NB: so tests are still failing. We're hitting something with a more complex scenario in jax -- so will need to drill down and understand why the backend_consistency_test fails.

**Comment by kratsg** [comment]
As noted in google/jax#936 -- setting jax to run in 64b mode fixes up the differences because of precision. The underlying issue is not understood (as to why the optimizer chooses 64b vs 32b or how it makes that distinction).

Additionally, it seems like the tests are timing out because it takes too long to respond.

**Comment by lukasheinrich** [comment]
what's missing here?

**Comment by matthewfeickert** [comment]
> what's missing here?

@lukasheinrich Hey sorry this fell off again. I think the only things that are missing is to get the optimizer setup correctly, and then get the backend consistency tests passing.

**Comment by lukasheinrich** [comment]
@matthewfeickert would also be great if we could test whether the  autodiff part / optimiizer works with this backend.

**Comment by matthewfeickert** [comment]
> would also be great if we could test whether the autodiff part / optimiizer works with this backend.

Yes, that's what I meant by "optimizer setup correctly" &mdash; sorry that wasn't clear.

**Comment by lukasheinrich** [comment]
@matthewfeickert what's the status here?

**Comment by matthewfeickert** [comment]
> @matthewfeickert what's the status here?

Now rebased off `master` so that it is up to date, but still the same status as before. I haven't been giving it enough time to do the final bits. I'm not sure if that will change in the next week, so anyone else should feel free to just work on it without worrying about stepping on anything I've got going.

**Comment by lukasheinrich** [comment]
@matthewfeickert I added a autodfiff based `jax_optimizer` that seems to work fine. It's still surprisingly slow, but it works

File: src/pyhf/optimize/opt_jax.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
+"""PyTorch Optimizer Backend."""
+
+from .. import get_backend, default_backend
+from ..tensor.common import _TensorViewer
+from .autodiff import AutoDiffOptimizerMixin
+import jax
+
+
+class jax_optimizer(AutoDiffOptimizerMixin):
+    """PyTorch Optimizer Backend."""
```
Comment (matthewfeickert) [inline src/pyhf/optimize/opt_jax.py]: Need to change the docstring name

**Comment by lukasheinrich** [comment]
ok jax is now the fastest hypothesis tester (after paying a one time cost for JIT'ing)

screenshot

**Pull request merged**